lib/rust/parser/src/syntax/tree.rs

```diff
@@ -755,6 +757,50 @@ impl<'s> span::Builder<'s> for OperatorDelimitedTree<'s> {
 
 
 
+// ================
+// === Warnings ===
+// ================
+
+/// Warnings applicable to a [`Tree`].
+pub type Warnings = ColdVec<Warning>;
+
+
+// === Warning ===
+
+/// A warning associated with a [`Tree`].
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Reflect, Deserialize)]
+#[allow(missing_copy_implementations)] // Future warnings may have attached information.
+pub struct Warning {
+    id: u32,
+}
+
+impl Warning {
+    /// Spacing is inconsistent with effective operator precedence.
+    pub fn inconsistent_spacing() -> Self {
+        Self { id: WarningId::InconsistentSpacing as u32 }
+    }
+
+    /// Return a description of the warning.
+    pub fn message(&self) -> Cow<'static, str> {
+        WARNINGS[self.id as usize].into()
+    }
+}
+
+#[repr(u32)]
+#[derive(Debug)]
+enum WarningId {
+    InconsistentSpacing,
+    #[allow(non_camel_case_types)]
+    NUM_WARNINGS,
+}
+
+/// Template strings for printing warnings.
+// These must be defined in the same order as the [`WarningId`] variants.
+pub const WARNINGS: [&str; WarningId::NUM_WARNINGS as usize] =
+    ["Spacing is inconsistent with operator precedence"];
```

Comment:
How about a function with a single match? It would not require `NUM_WARNINGS`, and is easier to read (no longer need to care about proper order), and - [if written properly](https://www.reddit.com/r/rust/comments/14zhb0s/assembly_examples_of_missed_rust_compiler/) - will optimize to array of &str anyway.

Reply:
The idea is to send the table across the FFI boundary, and then stringify warnings in the GC'd languages. This avoids FFI calls, encoding conversion, and allocations when emitting warnings. In the future we might move the warning table to a JSON file (which could be passed to some codegen) to further optimize this process and also make the messages easier to edit (or even localize).